### ISSUE TITLE
Can't get same scores as reported in paper using pytorch implementation

### ISSUE THREAD

**Issue by marzi-heidari** [computationalmedia/semstyle#9]
I used pytorch implementation for training my own models of img_to_txt and seq_to_txt but after 50 epoch of training the best BLEU score I can get for COCO_STYLE is 0.582 (for METEOR,ROUGE_L,CIDEr,SPICE best I could get are 0.192,0.445,0.633,0.134 respectively).
I know you used Theano implementation for the paper and using it probably gives better scores but the gap between scores of two implementations is pretty significant.Is there a way to get closer scores to theano implementation using pytorch implementation?

**Comment by almath123** [comment]
Thanks for letting me know. I have looked into it and can confirm there is a performance gap when COCO_STYLE is used. I am trying to work out why.

**Comment by annabeth-h** [comment]
Did you have any progress to a solution?

**Comment by marzi-heidari** [comment]
Did you have any progress to a solution?

**Comment by almath123** [comment]
No not yet. I unfortunately don't have a lot of time to work on this.

**Comment by marzi-heidari** [comment]
You don't have any clue where the problem might be? I want to try to solve it as a personal challenge. I've read your paper and I believe the code matches the descriptions in the experiment part. I don' have any experience with Theano and did not check the Theano implementation. Can you do me two favors and tell what part of the Pytorch implementation is different from Theano and provide a brief instruction of how to use Theano implementation to train a model from scratch. I try the instructions that you are provided already but I believe there are .txt and .json files that are needed and not present in the repository.

**Comment by Ruhallah93** [comment]
Hi @almath123 
I've this problem too
would you please solve it sooner?

**Comment by marzi-heidari** [comment]
I'm reaching my deadline soon and really need this problem to be solved :/

**Comment by almath123** [comment]
For comparison purposes the results of all the models/baselines on the entire testset (as generated by the theano code) are available in this json file:
https://github.com/almath123/semstyle_examples/blob/master/all_joined.json
If you want to compare you can use the above, all the test images are from MSCOCO and their filenames/ids are clearly given in the json.

As for the problem with the pytorch code, it is still on my todo list but I only have weekends (which more often than not are too full). So far I have converted the term generator into a caption generator and ensure that it gets the CNN+RNN accuracy listed in the paper -- this will pin down if the problem is in the termgen or langgen. Preliminary results suggest the termgen has a problem -- not 100% sure. Some possibilities are: use of relu between CNN and RNN rather than tanh as in the paper, dropout of 0.5 seems large and I need to confirm this was used, some decoding tricks (eg preventing duplicate words in the decoding process by dropping words out of the categorical distribution and re-normalising), gradient clipping in the RNN (not sure if pytorch does it the same way as theano).

**Comment by marzi-heidari** [comment]
OK. Using the following settings, I've got 
`Bleu_1: 0.655
Bleu_2: 0.486
Bleu_3: 0.347
Bleu_4: 0.247
METEOR: 0.223
ROUGE_L: 0.492
CIDEr: 0.813
SPICE: 0.163`
witch is enough for me.
I just changed ImgEmb to:
```
class ImgEmb(nn.Module):
    def __init__(self, input_size, hidden_size):
        super(ImgEmb, self).__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.emb_drop = nn.Dropout(0.2)
        self.mlp = nn.Linear(input_size, hidden_size)
        self.tanh = nn.Tanh()
    
    def forward(self, input):
        res = self.tanh(self.mlp(self.emb_drop(input)))
        return res
```
decoder to:
```
class Decoder(nn.Module):
    def __init__(self, input_size, hidden_size, output_size, out_bias=None):
        super(Decoder, self).__init__()
        self.hidden_size = hidden_size
        self.input_size = input_size
        self.embedding = nn.Embedding(input_size, hidden_size)
        self.emb_drop = nn.Dropout(0.2)
        self.gru = nn.GRU(hidden_size, hidden_size, batch_first=True)
        self.gru_drop = nn.Dropout(0.2)
        self.mlp = nn.Linear(hidden_size, output_size)
        if out_bias is not None:
            out_bias_tensor = torch.tensor(out_bias, requires_grad=False)
            self.mlp.bias.data[:] = out_bias_tensor
        self.logsoftmax = nn.LogSoftmax(dim=2)
    
    def forward(self, input, hidden_in):
        emb = self.embedding(input)
        emb=self.emb_drop(emb)
        out, hidden = self.gru(emb, hidden_in)
        out = self.mlp(self.gru_drop(out))
        out = self.logsoftmax(out)
        return out, hidden
```
(Just chenged drop outs)
trainer to:
```
def build_trainers(enc, dec, loaded_state = None):
    learning_rate = 0.001
    lossfunc = nn.NLLLoss(ignore_index=0)
    enc_optim = optim.Adam(enc.parameters(), lr=learning_rate,weight_decay=1e-6)
    dec_optim = optim.Adam(dec.parameters(), lr=learning_rate,weight_decay=1e-6)
    if loaded_state is not None:
        enc_optim.load_state_dict(load_state['enc_optim'])
        dec_optim.load_state_dict(load_state['dec_optim'])
    return enc_optim, dec_optim, lossfunc
```
(Just adding weight decay from theano implementation)
and using gradian cliping for just decoder during train

**Comment by almath123** [comment]
Great ! I'm glad you fixed the issue. Thank you for sharing what you did it is a great help.
